pallets/aux/shift-session-manager/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+        type Hashing = BlakeTwo256;
+        type AccountId = AccountId;
+        type Lookup = IdentityLookup<Self::AccountId>;
+        type Header = Header;
+        type Event = ();
+        type BlockHashCount = BlockHashCount;
+        type Version = ();
+        type PalletInfo = PalletInfo;
+        type AccountData = ();
+        type OnNewAccount = ();
+        type OnKilledAccount = ();
+        type BaseCallFilter = frame_support::traits::Everything;
+        type SystemWeightInfo = ();
+        type BlockWeights = ();
+        type BlockLength = ();
+        type DbWeight = ();
+        type SS58Prefix = ();
+        type OnSetCode = ();
+        type MaxConsumers = frame_support::traits::ConstU32<16>;
+    }
+
+    parameter_types! {
+        pub const Period: u64 = 1;
+        pub const Offset: u64 = 0;
+    }
+
+    impl pallet_session::Config for TestRuntime {
+        type Event = ();
+        type ValidatorId = <Self as frame_system::Config>::AccountId;
+        type ValidatorIdOf = ConvertInto;
+        type ShouldEndSession = pallet_session::PeriodicSessions<Period, Offset>;
+        type NextSessionRotation = pallet_session::PeriodicSessions<Period, Offset>;
+        type SessionManager = ();
+        type SessionHandler = TestSessionHandler;
+        type Keys = UintAuthorityId;
+        type WeightInfo = ();
+    }
+
+    parameter_types! {
+        pub const SelectedFraction: Percent = Percent::from_percent(50); // taking approximately half
+    }
+
+    impl Config for TestRuntime {
+        type SelectedFraction = SelectedFraction;
+    }
+
+    pub struct TestSessionHandler;
+    impl pallet_session::SessionHandler<AccountId> for TestSessionHandler {
+        const KEY_TYPE_IDS: &'static [sp_runtime::KeyTypeId] = &[UintAuthorityId::ID];
+
+        fn on_genesis_session<Ks: sp_runtime::traits::OpaqueKeys>(_validators: &[(AccountId, Ks)]) {
+        }
+
+        fn on_new_session<Ks: sp_runtime::traits::OpaqueKeys>(
+            _: bool,
+            _: &[(AccountId, Ks)],
+            _: &[(AccountId, Ks)],
+        ) {
+        }
+
+        fn on_disabled(_: u32) {}
+    }
+
+    fn new_test_ext() -> TestExternalities {
+        let mut t = frame_system::GenesisConfig::default()
+            .build_storage::<TestRuntime>()
+            .unwrap();
+
+        let keys = vec![
+            (1, 1, UintAuthorityId(1)),
+            (2, 2, UintAuthorityId(2)),
+            (3, 3, UintAuthorityId(3)),
+            (4, 4, UintAuthorityId(4)),
+            (5, 5, UintAuthorityId(5)),
+        ];
+
+        BasicExternalities::execute_with_storage(&mut t, || {
+            for (ref k, ..) in &keys {
+                frame_system::Pallet::<TestRuntime>::inc_providers(k);
+            }
+        });
+
+        pallet_session::GenesisConfig::<TestRuntime> { keys }
+            .assimilate_storage(&mut t)
+            .unwrap();
+        TestExternalities::new(t)
+    }
+
+    #[test]
+    fn shift_session_manager_works() {
```

Comment:
I'm right that the main purpose of this pallet is to provide different validators for different sessions, and following some circular buffer of current set of validators?

Reply:
And what will (and should) happen if we add new validator?

Reply:
This is a deterministic algorithm to select a certain fraction of pre-defined set of authorities as block producers for the next epoch. Essentially, this is the same old PoA. It's a temporary session management implementation used with `Babe` until a proper authorities selection based on staking is introduced. Then it will be possible to add validators.

In this simplistic use case we do not expect any equivocation or anything of that sort, so I think we should just make sure that exactly one authority is selected for each epoch.